### ISSUE TITLE
Error while training model

### ISSUE THREAD

**Issue by rhlr** [thekevinscott/ml-classifier#3]
I was trying your tutorial from here:https://thekevinscott.com/image-classification-with-javascript/

Everything works fine until I call model.fit(). Below is the exception I get.
**Error: Error when checking input: expected flatten_Flatten1_input to have shape [,7,7,256], but got array with shape [2,224,224,3]**

If I change this line: tf.layers.flatten({inputShape: [7, 7, 256]}) to tf.layers.flatten({inputShape: [224, 224, 3]}), the browser crashes. Can you tell me what I'm doing wrong ?

**Comment by thekevinscott** [comment]
@rhlr do you have a repo I can check out?

On first glance, it looks like the images may be missing the [mobilenet prediction step here](https://github.com/thekevinscott/ml-classifier/blob/master/src/index.ts#L89), as that should convert them from `224` to `7`, though if you're leveraging `ml-classifier` that should be taken care of automatically, in which case I'm not really sure.

If you could share a repo with this happening I'd be happy to take a further look.

**Comment by ymarcelo** [comment]
Hello, its an exelent tutorial this one here: https://thekevinscott.com/image-classification-with-javascript/

however, i got the same error explained above, I get that we need to predict with mobilenet first or so using const pred = this.pretrainedModel.predict(processedImage); or the equivalent …

my question is, where or when exactly do we need to do that?

would it be somewhere here: before or after data.concat..

return tf.tidy(() => {
          const processedImage = loadAndProcessImage(loadedImage, pretrainedModel);
          if (data) {
            const newData = data.concat(processedImage);
            data.dispose();
            return newData;
          }

          return tf.keep(processedImage);
        });
      });
    });
  }

or is it somewhere else??  I am very new to this really, so if you can please be explicit i would apreciate it

thanks man! you are a savior just for sharing your knowledge!

**Comment by thekevinscott** [comment]
Thank you, @ymarcelo!

You would want to predict with mobilenet after the call to `loadAndProcessImage`. I believe the code you're referring to is from the article you linked - in which case, I believe there is a bug. It should be:

```
return tf.tidy(() => {
          const processedImage = loadAndProcessImage(loadedImage);
          const prediction = pretrainedModel.predict(processedImage);

          if (data) {
            const newData = data.concat(prediction);
            data.dispose();
            return newData;
          }

          return tf.keep(prediction);
        });
```

[In this repo, the relevant line is here](https://github.com/thekevinscott/ml-classifier/blob/master/src/index.ts#L89).

Conceptually, the process looks like this:

1. Prepare the image - get the pixels, crop it, resize it to 224x224 (which is what mobilenet expects) and prepare the tensor
2. Push that tensor through mobilenet, and get a resulting 7x7x256 tensor (I believe those are the dimensions, I may be wrong)
3. Push THAT tensor through your custom training code

So, to answer your question - you want to call `pretrainedModel.predict` AFTER you've transformed the original image into a 224x224 image, but BEFORE you begin training your model.

Does that make sense?

**Comment by thekevinscott** [comment]
And thanks for bringing this to my attention - I'll try and get the article updated soon!

**Comment by ymarcelo** [comment]
Thank you so much my brother.
Actually, your explanation made it all clear now! Before this, I didnt really fully understand transfer learning. I mean, I still dont, but now I have a better picture of what goes on  in each step.

I was able to find what I believe is another bug in the tutorial. But this time, thanks to my new insight, provided by you, I was able to copy paste code from other parts of the tutorial and made it worked! 

So this part right here would break right in: console.log('Predicted Label', predict(model, loadedImage));:

```
function makePrediction(pretrainedModel, image, expectedLabel) {
  loadImage(image).then(loadedImage => {
    return loadAndProcessImage(loadedImage, pretrainedModel);
  }).then(loadedImage => {
    console.log('Expected Label', expectedLabel);
    console.log('Predicted Label', predict(model, loadedImage));
    loadedImage.dispose();
  });
}
```

So I changed it to this:

```
function makePrediction(pretrainedModel, model, image, expectedLabel) {
 		loadImage(image).then(loadedImage => {
    		return loadAndProcessImage(loadedImage, pretrainedModel);
  		}).then(loadedImage => {
    		
            //predict with mobilenet extracted layers
            const prepredict = pretrainedModel.predict(loadedImage);
            //put that tensor through new model
            const validating = model.predict(prepredict);
            //printing output tensors
            validating.print();
            //get the label value from the tensors
            const labelValidation = validating.as1D().argMax().dataSync()[0];
            //log it
            console.log('Expected Label', expectedLabel);
    		console.log('Predicted Label', labelValidation);
    		loadedImage.dispose();
  		});
	}

```
And it worked, I dont know if it was the best way to do it. But I guess that is one way to do it.
Anyways, thanks again. Cant wait to check out more of your tutorials!
Keep up with this outstanding work!

**Comment by thekevinscott** [comment]
Thank you for your help, @ymarcelo. I've updated the article with your updates (modified some of the names but mostly kept it intact. Thanks again!
